modules/infoserv/main.c

```diff
@@ -326,159 +289,250 @@ is_cmd_post(struct sourceinfo *si, int parc, char *parv[])
+		if (oper)
+			command_fail(si, fault_badparams, _("Syntax: OMOVE <from> <to>"));
+		else
+			command_fail(si, fault_badparams, _("Syntax: MOVE <from> <to>"));
 		return;
 	}
 
-	// search for it
-	MOWGLI_ITER_FOREACH(n, operlogon_info.head)
+	mowgli_list_t *list = (oper ? &operlogon_info : &logon_info);
+
+	if (!from_id || from_id > list->count)
 	{
-		o = n->data;
-		x++;
+		command_fail(si, fault_nosuch_target, _("The specified entry does not exist."));
+		return;
+	}
 
-		if (x == id)
-		{
-			logcommand(si, CMDLOG_ADMIN, "INFO:ODEL: \2%s\2, \2%s\2", o->subject, o->story);
+	if (!to_id || to_id > list->count)
+	{
+		command_fail(si, fault_nosuch_target, _("The target position does not exist."));
+		return;
+	}
 
-			mowgli_node_delete(n, &operlogon_info);
+	if (from_id == to_id)
+	{
+		command_fail(si, fault_nochange, _("You must specify two different positions."));
+		return;
+	}
 
-			strshare_unref(o->nick);
-			sfree(o->subject);
-			sfree(o->story);
-			sfree(o);
+	mowgli_node_t *from_node = NULL, *to_node = NULL, *n;
+	struct logoninfo *entry = NULL;
 
-			command_success_nodata(si, _("Deleted entry %u from oper logon info."), id);
-			return;
+	unsigned int pos = 0;
+	MOWGLI_ITER_FOREACH(n, list->head)
+	{
+		// 1-based indexing, reflecting what we display in LIST output
+		pos++;
+
+		if (!from_node && pos == from_id)
+		{
+			// We'll be removing this node from the list and adding it before the
+			// item currently at the target position. We need to skip this node
+			// in the count in case the target position is after it, however, as
+			// we're going to be removing it.
+			pos--;
+			from_node = n;
+			entry = n->data;
 		}
+
+		if (!to_node && pos == to_id)
+			to_node = n;
+
+		if (from_node && to_node)
+			break;
+	}
+
+	// this should never happen
+	// note that to_node can be NULL if we're asked to move an item to the end of the list;
+	// mowgli_node_insert_before will do the right thing in that case
+	if (!from_node || from_node == to_node)
+	{
+		if (!from_node)
+			slog(LG_ERROR, "%s: failed to find node %u in list (BUG)", MOWGLI_FUNC_NAME, from_id);
+		else
+			slog(LG_ERROR, "%s: trying to move node %u to position %u but found the same node (BUG)", MOWGLI_FUNC_NAME, from_id, to_id);
+
+		command_fail(si, fault_internalerror, _("There was an error modifying the list. This is a bug in services."));
+		return;
 	}
 
-	command_fail(si, fault_nosuch_target, _("Entry \2%u\2 not found in oper logon info."), id);
+	mowgli_node_delete(from_node, list);
+	mowgli_node_add_before(entry, from_node, list, to_node);
+
+	if (oper)
+		command_success_nodata(si, _("Oper logon info entry \2%u\2 moved to position \2%u\2."), from_id, to_id);
+	else
+		command_success_nodata(si, _("Logon info entry \2%u\2 moved to position \2%u\2."), from_id, to_id);
+	logcommand(si, CMDLOG_ADMIN, "INFO:%s: \2%u\2 -> \2%u\2: \2%s\2, \2%s\2", oper ? "OMOVE" : "MOVE", from_id, to_id, entry->subject, entry->story);
```

Comment:
Is it worth logging the notice contents (`entry->story`) here?

And how about we log the subject first, a la
```C
logcommand(si, CMDLOG_ADMIN, "INFO:%s: \2%s\2: \2%u\2 -> \2%u\2", oper ? "OMOVE" : "MOVE", entry->subject, from_id, to_id);
```

Reply:
I've mostly tried to keep this analogous to the existing log messages for `DEL`/`ODEL`, keeping in mind that the subject alone is not enough to identify an entry. (In fact, the only reliable way to identify an entry in the presence of complete duplicates would be its position, but that isn't very suitable as an identifier given it shifts around when entries are moved/deleted.)

Generally, though, I'd agree that it would be ideal if we had some sort of identifier that we could log first (before the `from -> to` part of the log message).

(Whether it'd be a good idea or not, requiring the subject to be unique (and thus allowing it to serve as an identifier) would break compatibility with existing setups that have multiple entries with the same subject (due to using it as more of a "category" field, perhaps).)

Reply:
Fair enough.